### PR TITLE
[ONNX] Add Support on 0d tensor Broadcast

### LINKED ISSUE #86789: ONNX export ShapeInferenceError with 0-sized tensors
### 🐛 Describe the bug

```python
from io import BytesIO
import torch
import torch.onnx

def fn(x, y):
    a = torch.add(x, y)
    b = torch.mul(y, y)
    return a + b

inputs = [torch.ones(1), torch.ones(0)]
scripted = torch.jit.trace(fn, inputs)
onnx_bytes = BytesIO()
torch.onnx.export(scripted,
                    inputs,
                    onnx_bytes,
                    opset_version=14,
                    input_names = ["x", "y"],
                    output_names = ["output"])
```

Outputs:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/torch/onnx/utils.py", line 1655, in _export
    _C._check_onnx_proto(proto, full_check=True)
RuntimeError: [ShapeInferenceError] (op_type:Add, node name: /Add_1): [ShapeInferenceError] Inferred shape and existing shape differ in dimension 0: (0) vs (1)
```

### Versions

nightly

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87212
* __->__ #87211

I am not sure if this will break things ...

Although 0d tensor is an undefined behavior in ONNX spec, I did some experiments and found that ONNX shape inference actually provides 0d as inference from 0d and 1d Op calculations, and the bug happened in Broadcast function. But still, if this breaks things really bad, I think we can put 0d tensor handling on hold, as it's not very common usage on models?

### CODE REVIEW

**Comment by AllenTiTaiWang** [comment]
cc @jcwchen

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 1]
```diff
@@ -703,15 +703,21 @@ std::vector<::c10::ShapeSymbol> Broadcast(
     if (is_static_0 && is_static_1) {
       int64_t static_0_sz = ss_shape_0.static_size();
       int64_t static_1_sz = ss_shape_1.static_size();
-      final_shape[rank_max - 1 - idx] = ::c10::ShapeSymbol::fromStaticSize(
-          std::max(static_0_sz, static_1_sz));
+      // condition for corner case of 0d tensor
+      // 0d tensor with 1d tensor would give us 0d tensor
+      if (std::min(static_0_sz, static_1_sz) == 0) {
+        final_shape[rank_max - 1 - idx] = ::c10::ShapeSymbol::fromStaticSize(
```
Comment (jcwchen) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:709]: nit: could we use shape_idx globally to represent `rank_max - 1 - idx` for simplicity?
```
size_t shape_idx = rank_max - 1 - idx;
```

**Review by jcwchen — APPROVED** [review]
Thanks for fixing this. With this PR, PyTorch, Exporter, ONNX shape inference should be all align with the case (0d, 1d) -> 0d.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 1 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:709]: Done

**Comment by AllenTiTaiWang** [comment]
@BowenBao PTAL

**Review by BowenBao — APPROVED** [review]
LGTM. If you want, you can add a unit test in `test_pytorch_onnx_shape_inference.py`

**Review by jcwchen — APPROVED** [review]

**Pull request closed without merging**